File: C++/Relationship.cpp
// Class relationships in C++ (C++17)
//
//   Association  - "uses-a"  (independent lifetimes, passed by reference)
//   Aggregation  - "has-a"   (whole holds non-owning refs to parts)
//   Composition  - "owns-a"  (whole owns parts, shared lifetime)

#include <iostream>
#include <memory>
#include <string>
#include <utility>
#include <vector>

// ------------------------------------------------------------
// Association
// ------------------------------------------------------------
class Student {
public:
    std::string name;
    explicit Student(std::string name) : name(std::move(name)) {}
};

class Payment {
public:
    double amount;
    explicit Payment(double amount) : amount(amount) {}
};

class Teacher {
public:
    void teaches(const std::vector<Student*>& students) const {
        for (const auto* student : students) {
            std::cout << "Teaching " << student->name << '\n';
        }
    }

    void getSalary(const Payment& payment) const {
        std::cout << "Teacher received " << payment.amount << '\n';
    }
};

// ------------------------------------------------------------
// Aggregation (students outlive the course)
// ------------------------------------------------------------
class LLDCourse {
    std::vector<Student*> students_;  // non-owning

public:
    void enroll(Student* student) { students_.push_back(student); }
    std::size_t size() const { return students_.size(); }
};

// ------------------------------------------------------------
// Composition (directory owns its children)
// ------------------------------------------------------------
struct IFileSystemNode {
    virtual ~IFileSystemNode() = default;
    virtual std::string name() const = 0;
};

class File : public IFileSystemNode {
    std::string name_;
    std::string meta_;

public:
    File(std::string name, std::string meta)
        : name_(std::move(name)), meta_(std::move(meta)) {}

    std::string name() const override { return name_; }
};

class Directory : public IFileSystemNode {
    std::string name_;
    std::vector<std::unique_ptr<IFileSystemNode>> children_;  // owned

public:
    explicit Directory(std::string name) : name_(std::move(name)) {}

    std::string name() const override { return name_; }

    void addChild(const std::string& name, const std::string& meta) {
        children_.push_back(std::make_unique<File>(name, meta));
    }

    const std::vector<std::unique_ptr<IFileSystemNode>>& children() const {
        return children_;
    }
};

int main() {
    Student neha("Neha");
    Student rahul("Rahul");
    std::vector<Student*> students{&neha, &rahul};

    Teacher teacher;
    teacher.teaches(students);
    teacher.getSalary(Payment(50000));

    LLDCourse course;
    for (auto* student : students) course.enroll(student);
    std::cout << "Course has " << course.size() << " students\n";

    Directory root("root");
    root.addChild("main.cpp", "text/x-c++");
    std::cout << root.name() << " contains:";
    for (const auto& child : root.children()) {
        std::cout << ' ' << child->name();
    }
    std::cout << '\n';

    return 0;
}

File: Python/Relationship.py
"""
Class relationships in Python.

    Association  - "uses-a"  (objects reference each other, independent lifetimes)
    Aggregation  - "has-a"   (whole references parts, parts outlive the whole)
    Composition  - "owns-a"  (whole creates/owns parts, shared lifetime)
"""

from __future__ import annotations

from abc import ABC, abstractmethod


# ------------------------------------------------------------
# Association
# ------------------------------------------------------------
class Student:
    def __init__(self, name: str) -> None:
        self.name = name


class Payment:
    def __init__(self, amount: float) -> None:
        self.amount = amount


class Teacher:
    def teaches(self, students: list[Student]) -> None:
        for student in students:
            print(f"Teaching {student.name}")

    def get_salary(self, payment: Payment) -> None:
        print(f"Teacher received {payment.amount}")


# ------------------------------------------------------------
# Aggregation
# ------------------------------------------------------------
class LLDCourse:
    def __init__(self) -> None:
        self.students: list[Student] = []  # students exist outside the course

    def enroll(self, student: Student) -> None:
        self.students.append(student)


# ------------------------------------------------------------
# Composition
# ------------------------------------------------------------
class IFileSystemNode(ABC):
    @abstractmethod
    def name(self) -> str:
        ...


class File(IFileSystemNode):
    def __init__(self, name: str, meta: str = "") -> None:
        self._name = name
        self._meta = meta

    def name(self) -> str:
        return self._name


class Directory(IFileSystemNode):
    def __init__(self, name: str) -> None:
        self._name = name
        self._children: list[IFileSystemNode] = []  # owned by this directory

    def name(self) -> str:
        return self._name

    def add_child(self, name: str, meta: str) -> None:
        self._children.append(File(name, meta))

    def children(self) -> list[IFileSystemNode]:
        return self._children


def main() -> None:
    students = [Student("Neha"), Student("Rahul")]

    teacher = Teacher()
    teacher.teaches(students)
    teacher.get_salary(Payment(50000))

    course = LLDCourse()
    for student in students:
        course.enroll(student)
    print(f"Course has {len(course.students)} students")

    root = Directory("root")
    root.add_child("main.py", "text/x-python")
    print(f"{root.name()} contains {[c.name() for c in root.children()]}")


if __name__ == "__main__":
    main()
